src/zenml/orchestrators/utils.py

```diff
@@ -134,13 +134,16 @@ def execute_step(
                 step_name = json.loads(
                     execution_info.exec_properties[step_name_param]
                 )
+                # TODO [MEDIUM]: find out which pipeline-run the cached
+                #  artifcats originate and log out the pipeline_run_id out here
                 logger.info(
-                    f"Using cached version of `{pipeline_step_name}` [`{step_name}`] "
-                    f"from pipeline_run_id `{pipeline_run_id}`.",
+                    f"Using cached version of `{pipeline_step_name}` "
+                    f"[`{step_name}`] from a previous pipeline run.",
```

Comment:
Im not sure where to get the old run either but @strickvl might have an idea

Reply:
At the moment we're benefiting from how `zenml` includes an `is_cached` property (generated from tfx's `status` property (which for a cached step would be `<ExecutionStatus.CACHED: 'cached'>`).

If we wanted to find out exactly how they make this determination ourselves, one way to do that would be to replicated their logic in determining whether a step is cached.

We have the `pipeline` object inside our `get_cache_status` method (which we grab from the `repository`. The current run is `pipeline.runs[-1]`, but we can get all the previous runs just with `pipeline.runs`.

We'd need to go through each run, and for each, extract the steps. For example: `pipeline.runs[-3].steps`. Then for whatever step we're on in the *current* pipeline run (i.e. with its own `id` value) we'd need to compare it against the previous execution of the step (with that same `id`).

The one thing I'm not fully sure about is where the calculation is made that a step (i.e. an 'execution' in tfx/ml_metadata speak) is the same or not, but possibly that would be another place where we could reach into.

The thing we'd be looking for would basically be the first execution of a particular step (with the same `id`) that did *not* have an execution status of `CACHED` but rather we'd want it to be `COMPLETED`. That seems like it should be doable, if a bit fiddly.

Reply:
@AlexejPenner Maybe that helps? If not, we can merge this PR with the current version with keeping the log to "`Using cached version of `{pipeline_step_name} [`{step_name}`]" (without the rest of the sentence)

Reply:
Couple of questions regarding your answer @strickvl :

Can we expect pipeline.runs to return a sorted list of runs from oldest to latest run?
Alternatively can we expect the IDs to increase with each run?
I need to have some reliable way to go for the `latest` run that contains a completed step. I think splicing teh timestamp out of the runname won't work as users can define their own run names without a timestamp in it.

Reply:
That's implemented by `tfx`, but as far as I can figure out from the code, yes you would get a sorted list from oldest to latest. At least, that's how we're currently able to get the same effect with `pipeline.runs`.

I'm not sure how the `id`s are constructed.

The 'solution' I suggested above feels doable, but definitely hacky / vulnerable to other things changing and breaking the behaviour.